### ISSUE TITLE
Defaults function empty lists error

### ISSUE THREAD

**Issue by horjulf** [hashicorp/terraform#28168]
Hi,

Using the new `defaults` function, lists can't be set empty as in variable defaults.
Unsure if this is by design, opening this issue for visibility/feedback.

### Terraform Version

```
v0.15.0-beta2
```

### Terraform Configuration Files

```terraform
variable "test" {
  type = object({
    list1 = optional(list(string))
    string1 = optional(string)
  })
  default = {}
}

locals {
  test = defaults(var.test, {
    list1 = []
    string1 = ""
  })
}

output "test" {
  value = local.test
}
```

### Debug Output

```
│ Invalid value for "defaults" parameter: .list1: invalid default value
│ for string: string required.
```

### Expected Behavior

Lists on the `defaults` function should be allowed to be empty, as it happens on variable default with the same type.

### Actual Behavior

Function error, requiring list to have content.

### Steps to Reproduce

1. `terraform init`
2. `terraform plan`

**Comment by alisdair** [comment]
Hi @horjulf, thanks for reporting this.

As far as I understand it, this is intentional behaviour. The `defaults` argument for a collection type is applied to each of its values, not to the collection as a whole. [From the docs](https://www.terraform.io/docs/language/functions/defaults.html):

> **Collection types** (`list`, `map`, and `set` types): Terraform will visit each of the collection elements in turn and apply defaults to them. In this case the default value is only a single value to be applied to _all_ elements of the collection, so it must have a type compatible with the collection's element type rather than with the collection type itself.

In the instance from your report, removing the default for `list1` will result in the output I think you expected:

```hcl
terraform {
  experiments = [module_variable_optional_attrs]
}

variable "test" {
  type = object({
    list1 = optional(list(string))
    string1 = optional(string)
  })
  default = {}
}

locals {
  test = defaults(var.test, {
    string1 = ""
  })
}

output "test" {
  value = local.test
}
```
```shellsession
$ terraform apply -auto-approve
╷
│ Warning: Experimental feature "module_variable_optional_attrs" is active
│
│   on main.tf line 2, in terraform:
│    2:   experiments = [module_variable_optional_attrs]
│
│ Experimental features are subject to breaking changes in future minor or patch releases, based on feedback.
│
│ If you have feedback on the design of this feature, please open a GitHub issue to discuss it.
╵

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

test = {
  "list1" = tolist([])
  "string1" = ""
}
```

Hopefully this lines up with what you're trying to do. I'd be interested to hear more about your use case here, if you're able to share it!

**Comment by horjulf** [comment]
Hi @alisdair,

I totally overlooked that and assumed the objects would be null, can confirm it defaults to an empty collection type if omitted from defaults.

Closing this.

Thanks !

**Comment by horjulf** [comment]
Just to add more into my use case, its related to #28170, I have modules with a lot of customization (60+ variables) that are used hundreds of times in a single state/run.

Using `defaults`, I'm merging most of those variables into object blocks, reducing the number of variables per module to workaround performance issues on module expansion.

**Comment by korinne** [comment]
Hi all, we recently released the [Terraform v1.3 alpha](https://releases.hashicorp.com/terraform/1.3.0-alpha20220608), which includes the ability to mark object type attributes as optional and set default values.

I'm following up with issues that have provided feedback on the previous experiment, and wanted to invite you all to try the alpha and provide any feedback on this updated design. You can learn more/add comments[ here](https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550). Thank you so much in advance, and hope you like the feature!
